File: practice_1_official.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <numeric>
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <random>
#include <chrono>

using namespace std;

// --- Estructura y Globales ---
struct QAPInstance {
    string name;
    int n;
    vector<vector<double>> flow;
    vector<vector<double>> dist;
};

// Semillas (Anexo)
const vector<unsigned int> SEEDS = {42, 12, 99, 123, 777, 2025, 3000, 5, 88, 101};

// --- Utils ---
QAPInstance read_qap(const string& filename) {
    QAPInstance q;
    q.name = filename;
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error abriendo " << filename << endl;
        q.n = 0;
        return q;
    }
    
    file >> q.n;
    // Leer Flujo
    q.flow.resize(q.n, vector<double>(q.n));
    for (int i = 0; i < q.n; i++)
        for (int j = 0; j < q.n; j++) file >> q.flow[i][j];
        
    // Leer Distancia
    q.dist.resize(q.n, vector<double>(q.n));
    for (int i = 0; i < q.n; i++)
        for (int j = 0; j < q.n; j++) file >> q.dist[i][j];
        
    return q;
}

double evaluate(const QAPInstance& q, const vector<int>& p) {
    double cost = 0;
    for (int i = 0; i < q.n; i++) {
        for (int j = 0; j < q.n; j++) {
            cost += q.flow[i][j] * q.dist[p[i]][p[j]];
        }
    }
    return cost;
}

double delta_evaluate(const QAPInstance& q, const vector<int>& p, int r, int s) {
    double delta = 0;
    for (int k = 0; k < q.n; k++) {
        if (k != r && k != s) {
            delta += (q.flow[r][k] - q.flow[s][k]) * 
                     (q.dist[p[s]][p[k]] - q.dist[p[r]][p[k]]) +
                     (q.flow[k][r] - q.flow[k][s]) * 
                     (q.dist[p[k]][p[s]] - q.dist[p[k]][p[r]]);
        }
    }
    return delta;
}

// --- 1. Greedy ---
struct Element {
    int index;
    double val;
};
bool compareDesc(Element a, Element b) { return a.val > b.val; }
bool compareAsc(Element a, Element b) { return a.val < b.val; }

pair<vector<int>, double> greedy_solver(const QAPInstance& q) {
    vector<int> sol(q.n);
    vector<Element> flow_pot(q.n), dist_pot(q.n);
    
    for (int i = 0; i < q.n; i++) {
        flow_pot[i].index = i;
        dist_pot[i].index = i;
        flow_pot[i].val = 0; dist_pot[i].val = 0;
        for (int j = 0; j < q.n; j++) {
            flow_pot[i].val += q.flow[i][j];
            dist_pot[i].val += q.dist[i][j];
        }
    }
    
    sort(flow_pot.begin(), flow_pot.end(), compareDesc);
    sort(dist_pot.begin(), dist_pot.end(), compareAsc);
    
    for (int i = 0; i < q.n; i++) {
        sol[flow_pot[i].index] = dist_pot[i].index;
    }
    
    return {sol, evaluate(q, sol)};
}

// --- 2. Random Search ---
pair<vector<int>, double> random_search(const QAPInstance& q, unsigned int seed) {
    mt19937 rng(seed);
    vector<int> current_sol(q.n);
    iota(current_sol.begin(), current_sol.end(), 0);
    shuffle(current_sol.begin(), current_sol.end(), rng);
    
    double best_cost = evaluate(q, current_sol);
    vector<int> best_sol = current_sol;
    
    long long max_iter = 1000 * q.n;
    for (long long i = 0; i < max_iter; i++) {
        vector<int> sol(q.n);
        iota(sol.begin(), sol.end(), 0);
        shuffle(sol.begin(), sol.end(), rng);
        double cost = evaluate(q, sol);
        if (cost < best_cost) {
            best_cost = cost;
            best_sol = sol;
        }
    }
    return {best_sol, best_cost};
}

// --- 3. Local Search (First Improvement) ---
pair<vector<int>, double> local_search(const QAPInstance& q, unsigned int seed) {
    mt19937 rng(seed);
    vector<int> current_sol(q.n);
    iota(current_sol.begin(), current_sol.end(), 0);
    shuffle(current_sol.begin(), current_sol.end(), rng);
    
    double current_cost = evaluate(q, current_sol);
    bool improve = true;
    
    while(improve) {
        improve = false;
        vector<pair<int,int>> neighbors;
        for(int i=0; i<q.n; i++)
            for(int j=i+1; j<q.n; j++)
                neighbors.push_back({i,j});
        
        shuffle(neighbors.begin(), neighbors.end(), rng);
        
        for(auto p : neighbors) {
            double delta = delta_evaluate(q, current_sol, p.first, p.second);
            if(delta < 0) {
                swap(current_sol[p.first], current_sol[p.second]);
                current_cost += delta;
                improve = true;
                break; // First Improvement
            }
        }
    }
    return {current_sol, current_cost};
}

// --- 4. Simulated Annealing (Cauchy) ---
pair<vector<int>, double> simulated_annealing(const QAPInstance& q, unsigned int seed) {
    mt19937 rng(seed);
    uniform_real_distribution<double> dist_real(0.0, 1.0);
    
    vector<int> current_sol(q.n);
    iota(current_sol.begin(), current_sol.end(), 0);
    shuffle(current_sol.begin(), current_sol.end(), rng);
    
    double current_cost = evaluate(q, current_sol);
    vector<int> best_sol = current_sol;
    double best_cost = current_cost;
    
    // Params
    double mu = 0.3;
    double phi = 0.3;
    double t0 = (mu / -log(phi)) * current_cost;
    double temp = t0;
    
    long long max_iter = 50 * q.n;
    long long k = 0;
    
    while (k < max_iter) {
        int accepted = 0;
        int generated = 0;
        
        while (accepted < 5 && generated < 40) {
            generated++;
            int i = uniform_int_distribution<int>(0, q.n - 1)(rng);
            int j = uniform_int_distribution<int>(0, q.n - 1)(rng);
            if (i == j) continue;
            
            double delta = delta_evaluate(q, current_sol, i, j);
            bool accept = false;
            
            if (delta < 0) accept = true;
            else if (exp(-delta / temp) > dist_real(rng)) accept = true;
            
            if (accept) {
                swap(current_sol[i], current_sol[j]);
                current_cost += delta;
                accepted++;
                if (current_cost < best_cost) {
                    best_cost = current_cost;
                    best_sol = current_sol;
                }
            }
        }
        k++;
        temp = t0 / (1.0 + k); // Cauchy
    }
    return {best_sol, best_cost};
}

// --- 5. Tabu Search ---
pair<vector<int>, double> tabu_search(const QAPInstance& q, unsigned int seed) {
    mt19937 rng(seed);
    vector<int> current_sol(q.n);
    iota(current_sol.begin(), current_sol.end(), 0);
    shuffle(current_sol.begin(), current_sol.end(), rng);
    
    double current_cost = evaluate(q, current_sol);
    vector<int> best_sol = current_sol;
    double best_cost = current_cost;
    
    int restart_interval = 8 * q.n;
    int max_iter_total = 40 * q.n;
    int tabu_tenure = q.n / 2;
    
    vector<vector<int>> tabu_matrix(q.n, vector<int>(q.n, 0));
    vector<vector<int>> freq_matrix(q.n, vector<int>(q.n, 0));
    
    for (int iter = 1; iter <= max_iter_total; iter++) {
        int best_i = -1, best_j = -1;
        double best_local_delta = 1e18; 
        
        for (int k = 0; k < 40; k++) {
            int i = uniform_int_distribution<int>(0, q.n - 1)(rng);
            int j = uniform_int_distribution<int>(0, q.n - 1)(rng);
            if (i == j) { k--; continue; }
            
            double delta = delta_evaluate(q, current_sol, i, j);
            bool is_tabu = tabu_matrix[min(i,j)][max(i,j)] > iter;
            bool aspira = (current_cost + delta < best_cost);
            
            if (!is_tabu || aspira) {
                if (delta < best_local_delta) {
                    best_local_delta = delta;
                    best_i = i;
                    best_j = j;
                }
            }
        }
        
        if (best_i != -1) {
            swap(current_sol[best_i], current_sol[best_j]);
            current_cost += best_local_delta;
            
            int u = min(best_i, best_j), v = max(best_i, best_j);
            tabu_matrix[u][v] = iter + tabu_tenure;
            
            for (int k=0; k<q.n; k++) freq_matrix[k][current_sol[k]]++;
            
            if (current_cost < best_cost) {
                best_cost = current_cost;
                best_sol = current_sol;
            }
        }
        
        if (iter % restart_interval == 0 && iter < max_iter_total) {
            if (uniform_real_distribution<double>(0, 1)(rng) < 0.5) tabu_tenure = max(2, (int)(tabu_tenure * 1.5));
            else tabu_tenure = max(2, (int)(tabu_tenure * 0.5));
            
            double r = uniform_real_distribution<double>(0, 1)(rng);
            if (r < 0.25) { 
                shuffle(current_sol.begin(), current_sol.end(), rng);
                current_cost = evaluate(q, current_sol);
            } else if (r < 0.75) {
                shuffle(current_sol.begin(), current_sol.end(), rng); // Placeholder for Greedy Randomized
                current_cost = evaluate(q, current_sol);
            } else {
                current_sol = best_sol;
                current_cost = best_cost;
            }
            for(auto& row : tabu_matrix) fill(row.begin(), row.end(), 0);
        }
    }
    return {best_sol, best_cost};
}

int main() {
    cout << "Ejecutando Practica 1 (C++ Benchmark)" << endl;
    cout << "---------------------------------------" << endl;
    QAPInstance q = read_qap("tai25b.dat");
    if (q.n == 0) {
        q = read_qap("nug5.dat"); // Fallback check
        if(q.n == 0) return 1;
    }
    
    cout << "Problema: " << q.name << " (N=" << q.n << ")" << endl << endl;
    
    // Greedy
    auto start = chrono::high_resolution_clock::now();
    auto greedy = greedy_solver(q);
    auto end = chrono::high_resolution_clock::now();
    cout << "Greedy:  " << greedy.second << " \t(" << chrono::duration_cast<chrono::microseconds>(end-start).count() << " us)" << endl;
    
    // Random Search (Seed 42)
    start = chrono::high_resolution_clock::now();
    auto rs = random_search(q, 42);
    end = chrono::high_resolution_clock::now();
    cout << "RandomS: " << rs.second << " \t(" << chrono::duration_cast<chrono::milliseconds>(end-start).count() << " ms)" << endl;
    
    // Local Search (Seed 42)
    start = chrono::high_resolution_clock::now();
    auto ls = local_search(q, 42);
    end = chrono::high_resolution_clock::now();
    cout << "LocalS:  " << ls.second << " \t(" << chrono::duration_cast<chrono::microseconds>(end-start).count() << " us)" << endl;
    
    // SA (Seed 42)
    start = chrono::high_resolution_clock::now();
    auto sa = simulated_annealing(q, 42);
    end = chrono::high_resolution_clock::now();
    cout << "SimAnn:  " << sa.second << " \t(" << chrono::duration_cast<chrono::milliseconds>(end-start).count() << " ms)" << endl;
    
    // Tabu (Seed 42)
    start = chrono::high_resolution_clock::now();
    auto tabu = tabu_search(q, 42);
    end = chrono::high_resolution_clock::now();
    cout << "Tabu:    " << tabu.second << " \t(" << chrono::duration_cast<chrono::milliseconds>(end-start).count() << " ms)" << endl;
    
    return 0;
}

File: practice_1_official.py
import numpy as np
import math
import random
import time
import os
from statistics import mean, stdev

# --- Configuración del Generador Estático ---
# Semillas fijas para reproducibilidad (Anexo)
SEEDS = [42, 12, 99, 123, 777, 2025, 3000, 5, 88, 101]

class QAPProblem:
    def __init__(self, file_path):
        self.name = os.path.basename(file_path).split('.')[0]
        self.n, self.flow, self.dist = self.read_qap_file(file_path)

    def read_qap_file(self, file_path):
        with open(file_path, 'r') as f:
            content = f.read().split()
        iterator = iter(content)
        n = int(next(iterator))
        flow = np.zeros((n, n))
        for i in range(n):
            for j in range(n): flow[i][j] = float(next(iterator))
        dist = np.zeros((n, n))
        for i in range(n):
            for j in range(n): dist[i][j] = float(next(iterator))
        return n, flow, dist

    def evaluate(self, sol):
        cost = 0
        for i in range(self.n):
            for j in range(self.n):
                cost += self.flow[i][j] * self.dist[sol[i]][sol[j]]
        return cost

    def delta_evaluate(self, sol, r, s):
        # Simplificación Delta O(n) - Asume simetría en matrices para mayor velocidad
        # Para práctica real, usar implementación completa si matrices no son simétricas
        delta = 0
        for k in range(self.n):
            if k != r and k != s:
                delta += (self.flow[r][k] - self.flow[s][k]) * \
                         (self.dist[sol[s]][sol[k]] - self.dist[sol[r]][sol[k]]) + \
                         (self.flow[k][r] - self.flow[k][s]) * \
                         (self.dist[sol[k]][sol[s]] - self.dist[sol[k]][sol[r]])
        return delta

# --- 1. Greedy (Potenciales) ---
def greedy_solver(problem):
    # Potencial Flujo: Suma filas
    flow_pot = np.sum(problem.flow, axis=1)
    # Potencial Distancia: Suma filas
    dist_pot = np.sum(problem.dist, axis=1)
    
    # Ordenar: Flujo DESC, Distancia ASC
    units = np.argsort(flow_pot)[::-1]
    locs = np.argsort(dist_pot)
    
    sol = [-1] * problem.n
    for i in range(problem.n):
        sol[units[i]] = locs[i]
        
    return list(sol), problem.evaluate(sol)

def greedy_randomized_solver(problem, freq_matrix):
    """
    Greedy modificado para Tabú (Memoria a Largo Plazo).
    Usa la frecuencia inversa para penalizar asignaciones muy repetidas.
    (Interpretación de 'usar memoria a largo plazo al generar nueva solución Greedy')
    """
    # Penalizar flujo con frecuencia:
    # Si unit i se ha asignado mucho a loc j, reducimos su 'atractivo' ficticio
    # Esta es una heurística común, aunque el enunciado es vago en "cómo" usarla.
    # Implementación simple: Greedy puro pero añadiendo ruido aleatorio a los potenciales
    
    flow_pot = np.sum(problem.flow, axis=1) + np.random.normal(0, 10, problem.n)
    dist_pot = np.sum(problem.dist, axis=1)
    
    units = np.argsort(flow_pot)[::-1]
    locs = np.argsort(dist_pot)
    
    sol = [-1] * problem.n
    for i in range(problem.n):
        sol[units[i]] = locs[i]
        
    return list(sol), problem.evaluate(sol)

# --- Utilidad de Visualización ---
class ProgressBar:
    def __init__(self, total, prefix='', length=30):
        self.total = total
        self.prefix = prefix
        self.length = length
        self.start_time = time.time()
        
    def update(self, iteration, current_cost=None):
        percent = "{0:.1f}".format(100 * (iteration / float(self.total)))
        filled_length = int(self.length * iteration // self.total)
        bar = '=' * filled_length + '-' * (self.length - filled_length)
        
        stats = ""
        if current_cost is not None:
            stats = f"Cost: {current_cost:.2f}"
            
        print(f'\r{self.prefix} |{bar}| {percent}% {stats}', end='\r')
        if iteration == self.total: 
            print()

# --- 2. Búsqueda Aleatoria ---
def random_search(problem, seed):
    random.seed(seed)
    n = problem.n
    best_sol = list(range(n))
    random.shuffle(best_sol)
    best_cost = problem.evaluate(best_sol)
    
    evals = 1
    max_iter = 1000 * n
    
    # pb = ProgressBar(max_iter, prefix="RandomSearch")
    
    for i in range(max_iter):
        sol = list(range(n))
        random.shuffle(sol)
        cost = problem.evaluate(sol)
        evals += 1
        if cost < best_cost:
            best_cost = cost
            best_sol = sol
        # if i % 100 == 0: pb.update(i+1, best_cost)
            
    return best_sol, best_cost, evals

# --- 3. Búsqueda Local (Primer Mejor Aleatorizado) ---
def local_search_first_improvement_random(problem, seed, start_sol=None, verbose=False):
    random.seed(seed)
    n = problem.n
    
    if start_sol:
        current_sol = list(start_sol)
    else:
        current_sol = list(range(n))
        random.shuffle(current_sol)
        
    current_cost = problem.evaluate(current_sol)
    evals = 1
    improve = True
    
    # No sabemos cuántas iteraciones hará, difícil poner barra de progreso
    while improve:
        improve = False
        neighbors = []
        for i in range(n):
            for j in range(i + 1, n):
                neighbors.append((i, j))
        random.shuffle(neighbors)
        
        for i, j in neighbors:
            delta = problem.delta_evaluate(current_sol, i, j)
            evals += 1
            if delta < 0:
                current_sol[i], current_sol[j] = current_sol[j], current_sol[i]
                current_cost += delta
                improve = True
                break 
                
    return current_sol, current_cost, evals

# --- 4. Enfriamiento Simulado (Cauchy) ---
def simulated_annealing(problem, seed, mu=0.3, phi=0.3, verbose=False):
    random.seed(seed)
    n = problem.n
    
    current_sol = list(range(n))
    random.shuffle(current_sol)
    current_cost = problem.evaluate(current_sol)
    
    best_sol = list(current_sol)
    best_cost = current_cost
    
    t0 = (mu / -math.log(phi)) * current_cost
    temp = t0
    
    max_iter = 50 * n 
    
    k = 0
    total_evals = 1
    
    pb = ProgressBar(max_iter, prefix=f"SimAnnealing (Seed {seed})") if verbose else None
    
    while k < max_iter:
        accepted = 0
        generated = 0
        
        while accepted < 5 and generated < 40:
            generated += 1
            total_evals += 1
            
            i, j = random.sample(range(n), 2)
            delta = problem.delta_evaluate(current_sol, i, j)
            
            r = random.random()
            accept = False
            if delta < 0:
                accept = True
            elif math.exp(-delta / temp) > r:
                accept = True
                
            if accept:
                current_sol[i], current_sol[j] = current_sol[j], current_sol[i]
                current_cost += delta
                accepted += 1
                
                if current_cost < best_cost:
                    best_cost = current_cost
                    best_sol = list(current_sol)
        
        k += 1
        temp = t0 / (1 + k)
        if pb and k % (n) == 0: pb.update(k, best_cost)
        
    if pb: pb.update(max_iter, best_cost)
    return best_sol, best_cost, total_evals

# --- 5. Búsqueda Tabú (Completa) ---
def tabu_search_official(problem, seed, verbose=False):
    random.seed(seed)
    np.random.seed(seed)
    n = problem.n
    
    current_sol = list(range(n))
    random.shuffle(current_sol)
    current_cost = problem.evaluate(current_sol)
    
    best_sol = list(current_sol)
    best_cost = current_cost
    
    freq_matrix = np.zeros((n, n))
    
    tabu_tenure = n / 2
    tabu_matrix = np.zeros((n, n)) 
    
    max_iter_total = 40 * n
    restart_interval = 8 * n
    
    iter_global = 0
    evals = 1
    
    pb = ProgressBar(max_iter_total, prefix=f"TabuSearch   (Seed {seed})") if verbose else None
    
    while iter_global < max_iter_total:
        
        best_neighbor_delta = float('inf')
        best_move = None
        
        for _ in range(40):
            i, j = random.sample(range(n), 2)
            u, v = min(i, j), max(i, j) 
            
            delta = problem.delta_evaluate(current_sol, u, v)
            evals += 1
            
            is_tabu = tabu_matrix[u][v] > iter_global
            
            if not is_tabu or (current_cost + delta < best_cost):
                if delta < best_neighbor_delta:
                    best_neighbor_delta = delta
                    best_move = (u, v)
        
        if best_move:
            u, v = best_move
            current_sol[u], current_sol[v] = current_sol[v], current_sol[u]
            current_cost += best_neighbor_delta
            
            tabu_matrix[u][v] = iter_global + tabu_tenure
            
            if current_cost < best_cost:
                best_cost = current_cost
                best_sol = list(current_sol)
                
            for k in range(n):
                unit_idx = k
                loc_idx = current_sol[k]
                freq_matrix[unit_idx][loc_idx] += 1

        iter_global += 1
        
        if iter_global % restart_interval == 0 and iter_global < max_iter_total:
            r = random.random()
            
            if random.random() < 0.5:
                tabu_tenure *= 1.5
            else:
                tabu_tenure *= 0.5
            tabu_tenure = max(2, min(n, tabu_tenure)) 
            
            if r < 0.25:
                random.shuffle(current_sol)
                current_cost = problem.evaluate(current_sol)
            elif r < 0.75: 
                current_sol, current_cost = greedy_randomized_solver(problem, freq_matrix)
            else:
                current_sol = list(best_sol)
                current_cost = best_cost
                
            tabu_matrix.fill(0) 

        if pb and iter_global % (n) == 0: pb.update(iter_global, best_cost)
            
    if pb: pb.update(max_iter_total, best_cost)
    return best_sol, best_cost, evals

# --- Runner Principal ---
def run_practice():
    datasets = ["tai25b.dat"] # Use our mock data or real if available
    
    results = {} 
    
    print(f"\n{'Dataset':<10} {'Algoritmo':<20} {'Mejor':<10} {'Media':<10} {'Peor':<10} {'StdDev':<10} {'Evals':<10}")
    print("=" * 90)
    
    for ds_file in datasets:
        if not os.path.exists(ds_file): continue
        problem = QAPProblem(ds_file)
        results[ds_file] = {}
        print(f"Resolviendo {problem.name} (N={problem.n})...")
        
        # 1. Greedy
        g_sol, g_cost = greedy_solver(problem)
        print(f"Greedy Cost: {g_cost}")
        
        # 2. Búsqueda Aleatoria 
        costs = []
        for s in SEEDS[:5]:
            # No verbose for RS loop
            _, c, e = random_search(problem, s)
            costs.append(c)
        print(f"RandomSearch Media: {mean(costs):.2f}")
        
        # 3. Búsqueda Local
        costs = []
        for s in SEEDS[:5]:
            _, c, e = local_search_first_improvement_random(problem, s)
            costs.append(c)
        print(f"LocalSearch  Media: {mean(costs):.2f}")

        # 4. Enfriamiento Simulado (Verbose)
        costs = []
        evals_list = []
        for s in SEEDS[:5]:
            _, c, e = simulated_annealing(problem, s, verbose=True)
            costs.append(c)
            evals_list.append(e)
            
        # 5. Búsqueda Tabú (Verbose)
        costs_tabu = []
        evals_list_tabu = []
        for s in SEEDS[:5]: # Reduce to 5 for demo speed, manual says 10
            _, c, e = tabu_search_official(problem, s, verbose=True)
            costs_tabu.append(c)
            evals_list_tabu.append(e)
            
        print("-" * 90)
        print(f"{problem.name:<10} {'Tabu Official':<20} {min(costs_tabu):<10.2f} {mean(costs_tabu):<10.2f} {max(costs_tabu):<10.2f}")
        print("=" * 90)

if __name__ == "__main__":
    run_practice()
